### PR TITLE
Bug 2094854: api-resource-collector: Page MachineConfigs when listing them and throw away the file contents

### PR DESCRIPTION
**Pull request by jhrozek**
- Upgrade the MCO dependency to a version that handles Ignition 3.2.0
We need to parse Ignition files in api-resource collector using helper utilities from MCO, but in the meantime, the Ignition standard bumped the spec to 3.2.0 which our vendored MCO dependency didn't have. So let's bump the dependency to a version that supports Ignition 3.2.0

- api-resource-collector: Use paging and filter file contents when listing MachineConfigs
We have seen reports of the api-resource-collector being killed when used for the PCI-DSS standard. Turns out that the PCI-DSS content lists all MachineConfigs and on very large clusters with many MachineConfigs, this can amount to tens of megabytes that the api-resource-collector would then keep in memory.
Because we don't intend to inspect file contents of MCs (this is what the OS-level checks are for), let's strip the file contents out of MCs. In order to keep everything in memory before the files are filtered out, this patch also uses paging using the "limit" and "continue" query parameters of k8s API as described here: https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/list-meta/

- scan: Bump the memory limit of the api-resource collector to 200Mi
Even with memory optimizations, the 100Mi limit might be too strict to list all objects in a cluster. Let's bump the limit to 200Mi.

### CODE REVIEW

**Comment by jhrozek** [comment]
/hold
for QE review

**Comment by jhrozek** [comment]
[genmc.tar.gz](https://github.com/ComplianceAsCode/compliance-operator/files/8925740/genmc.tar.gz)

Attaching a script that I used to create a number of MCs to stress test the api-resource-collector. I used the script like this:
```
python genmc.py --in-file=/usr/share/dict/linux.words --num-files=100 --start-num=0 --create
```
the file to use as input is up to you of course. The MCs are not assigned to any pool to avoid triggering a reboot.
To find out the size of MCs in your cluster, you can fetch the MCs to a directory, run `du -sh` on it, or use something like `pv`:
```
oc get mc -oyaml | pv
```

**Comment by jhrozek** [comment]
Also please note that currently the only profile that triggers the bug is PCI-DSS. So to reproduce the issue, you'd create at least approx 50MB of MCs (or more..), then create a scan that scans the platform (k8s). Without the patch, the `api-resource-collector` pod would be `OOMKill`-ed, the memory usage should be much better with the patch.

File: cmd/manager/scap.go [hunk 1]
```diff
@@ -336,25 +355,132 @@ func (c *scapContentDataStream) getExtendedProfileFromTailoring(ds *xmlquery.Nod
...
+	return buf, nil
+}
+
+func filterMcList(mcListIn *mcfgv1.MachineConfigList) (*mcfgv1.MachineConfigList, error) {
+	mcfgListNoFiles := mcfgv1.MachineConfigList{}
+	mcfgListNoFiles.TypeMeta = mcListIn.TypeMeta
+	mcfgListNoFiles.ListMeta = mcListIn.ListMeta
+
+	for i := 0; i < len(mcListIn.Items); i++ {
+		mcCopy := mcListIn.Items[i].DeepCopy()
```
[the remark this answers is not in the document]
Comment (jhrozek) [reply cmd/manager/scap.go]: I think you're right. At least I cannot think of any other reason than "deepcopy looks safer" and the memory consumption went down quite a bit.

**Comment by mrogers950** [comment]
> [genmc.tar.gz](https://github.com/ComplianceAsCode/compliance-operator/files/8925740/genmc.tar.gz)
> 
> Attaching a script that I used to create a number of MCs to stress test the api-resource-collector. I used the script like this:
> 
> ```
> python genmc.py --in-file=/usr/share/dict/linux.words --num-files=100 --start-num=0 --create
> ```
> 
> the file to use as input is up to you of course. The MCs are not assigned to any pool to avoid triggering a reboot. To find out the size of MCs in your cluster, you can fetch the MCs to a directory, run `du -sh` on it, or use something like `pv`:
> 
> ```
> oc get mc -oyaml | pv
> ```

BTW, I'm not opposed to a testing side-script like this going in the tree under tests/extras or similar. It would be better there than lost in PR history.

**Comment by jhrozek** [comment]
On Fri, Jun 17, 2022 at 09:55:48AM -0700, Matt Rogers wrote:
> > [genmc.tar.gz](https://github.com/ComplianceAsCode/compliance-operator/files/8925740/genmc.tar.gz)
> > 
> > Attaching a script that I used to create a number of MCs to stress test the api-resource-collector. I used the script like this:
> > 
> > ```
> > python genmc.py --in-file=/usr/share/dict/linux.words --num-files=100 --start-num=0 --create
> > ```
> > 
> > the file to use as input is up to you of course. The MCs are not assigned to any pool to avoid triggering a reboot. To find out the size of MCs in your cluster, you can fetch the MCs to a directory, run `du -sh` on it, or use something like `pv`:
> > 
> > ```
> > oc get mc -oyaml | pv
> > ```
> 
> BTW, I'm not opposed to a testing side-script like this going in the tree under tests/extras or similar. It would be better there than lost in PR history.

Sure, good idea, I'll adjust the PR. Initially I added the script only as
attachment for QE's sake.

btw I didn't write an e2e test because I thought priming the MCs
would be too time-consuming. But at the same time I rarely run /all/ e2e
tests locally, so I can write an e2e test as well if you think it would
be better..

**Comment by JAORMX** [comment]
@jhrozek @mrogers950 what about writing a unit test that would test a significant amount of MCs?

**Comment by JAORMX** [comment]
> > @jhrozek @mrogers950 what about writing a unit test that would test a significant amount of MCs?
> 
> Sure, but what would that test? With e2e tests we'd know that the pod can finish given the default memory limits, but the unit tests are unbound when it comes to memory limits.

@jhrozek aren't the unit tests also running on a pod in a kube cluster? what's the resource limit set there? (We could probably find out by checking the `openshift/release` settings.

**Review by rhmdnd — COMMENTED** [review]
Only minor suggestions inline. Otherwise this looks good.

**Comment by jhrozek** [comment]
On Thu, Jun 23, 2022 at 02:17:45AM -0700, Juan Antonio Osorio wrote:
> > > @jhrozek @mrogers950 what about writing a unit test that would test a significant amount of MCs?
> > 
> > Sure, but what would that test? With e2e tests we'd know that the pod can finish given the default memory limits, but the unit tests are unbound when it comes to memory limits.
> 
> @jhrozek aren't the unit tests also running on a pod in a kube cluster? what's the resource limit set there? (We could probably find out by checking the `openshift/release` settings.

yes, with a 400Mi limit.

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,11 +20,22 @@ Versioning](https://semver.org/spec/v2.0.0.html).
   about non-compliant pods, would trigger events with warnings.
   See this [bug](https://bugzilla.redhat.com/show_bug.cgi?id=2088202) for more
   information.
+- The `api-resourcei-collector` no longer fetches and keeps the whole MachineConfig
+  objects when the content needs to examine and filter them, but instead filters
+  out the file contents of the MachineConfigs. This addressses
+  [bugs](https://bugzilla.redhat.com/show_bug.cgi?id=2094854) where especially
```
[the remark this answers is not in the document]
Comment (jhrozek) [reply CHANGELOG.md:22]: I pruned the BZs and marked one as a duplicate of this one. We still have one more, but in that case it's the openscap scanner that is OOM-ing.

[the remark this answers is not in the document]
Comment (jhrozek) [reply CHANGELOG.md:22]: I also made sure that the cases attached to the respective bugzillas really talk about the crash the BZ addresses and not the other one.

File: cmd/manager/scap.go [hunk 3]
```diff
@@ -336,25 +355,132 @@ func (c *scapContentDataStream) getExtendedProfileFromTailoring(ds *xmlquery.Nod
...
+		}
+	}
+
+	jsonSerializer := runtimejson.NewSerializerWithOptions(runtimejson.DefaultMetaFactory,
+		rfClients.scheme,
+		rfClients.scheme,
+		runtimejson.SerializerOptions{Pretty: true})
+	buf := &bufCloser{&bytes.Buffer{}}
+	if err := jsonSerializer.Encode(&mcfgListNoFiles, buf); err != nil {
+		return nil, fmt.Errorf("failed to list serialize MC list: %w", err)
```
[the remark this answers is not in the document]
Comment (jhrozek) [reply cmd/manager/scap.go:446]: good catch, will fix

**Comment by jhrozek** [comment]
@rhmdnd thanks for the careful review, I hope I've addressed your concerns.

**Comment by rhmdnd** [comment]
Looks good to me. Thanks @jhrozek!

/lgtm

**Comment by rhmdnd** [comment]
Removing the lgtm label until we get a clean test run.

**Comment by jhrozek** [comment]
I just rebased the changelog conflict. No other changes. Please re-add LGTM. @rhmdnd

**Pull request merged**